prometheus/doc.go

```diff
@@ -61,51 +62,135 @@
+//
+// Custom Collectors and constant Metrics
+//
+// While you could create your own implementations of Metric, most likely you
+// will only ever implement the Collector interface on your own. At a first
+// glance, a custom Collector seems handy to bundle Metrics for common
+// registration (with the prime example of the different metric vectors above,
+// which bundle all the metrics of the same name but with different labels).
+//
+// There is a more involved use-case, too: If you already have metrics
+// availably, created outside of the Prometheus context, you don't need the
+// interface of the various Metric types. You essentially want to mirror the
+// existing numbers into Prometheus Metrics during collection. An own
+// implementation of the Collector interface is perfect for that. You can create
+// Metric instances “on the fly” using NewConstMetric, NewConstHistogram, and
+// NewConstSummary (and their respective Must… versions). That will happen in
+// the Collect method, while the Describe method has to return separate Desc
+// instances, the very same you later use to create the constant “throw-away”
+// metrics. NewDesc comes in handy to create those Desc instances.
+//
+// The Collector example illustrates the use-case. You can also look at the
+// source code of the processCollector (mirroring process metrics), the
+// goCollector (mirroring Go metrics), or the exvarCollector (mirroring expvar
+// metrics) as examples that are used in this package itself.
+//
+// If you just need to call a function to get a single float value to collect as
+// a metric, GaugeFunc, CounterFunc, or UntypedFunc might be interesting
+// shortcuts.
+//
+// Advanced Uses of the Registry
+//
+// While MustRegister is the by far most common way of registering a Collector,
+// sometimes you might want to handle the errors the registration might
+// cause. As suggested by the name, MustRegister panics if an error occurs. With
+// the Register function, the error is returned and can be handled.
+//
+// An error is returned if the registered Collector is incompatible or
+// inconsistent with already registered metrics. The registry aims for
+// consistency of the collected metrics according to the Prometheus data
+// model. Inconsistencies are ideally detected at registration time, not at
+// collect time. The former will usually be detected at start-up time of a
+// program, while the latter will only happen at scrape time, possibly not even
+// on the first scrape if the inconsistency only becomes relevant later. That is
+// the main reason why a Collector and a Metric have to describe themselves to
+// the registry.
+//
+// A particular error case to handle is if a Collector has already be registered
+// (or a different Collector that collects the same metrics). Registration fails
+// in that case, but you can detect from the kind of error what has happened. A
+// common pattern is that you want to register the new Collector if an equal
```

Comment:
I found ~30 uses on GitHub. I didn't evaluate every single case for how legitimate they are. My own legitimate use case is in `InstrumentHandler` (which currently allows to instrument different handlers with the same metric).

In the spirit of reducing top-level identifiers, I'll remove the `...OrGet` for custom registries, see other comment thread.